keps/140-scenario-based-simulation/README.md

```diff
@@ -0,0 +1,410 @@
+  // Filter has each filter plugin’s result.
+  Filter            map[NodeName]map[PluginName]string
+  // Score has each score plugin’s score.
+  Score             map[NodeName]map[PluginName]ScenarioPluginsScoreResult
+}
+
+type ScenarioPluginsScoreResult struct {
+  // RawScore has the score from Score method of Score plugins.
+  RawScore             int64
+  // NormalizedScore has the score calculated by NormalizeScore method of Score plugins.
+  NormalizedScore      int64
+  // FinalScore has score plugin’s final score calculated by normalizing with NormalizedScore and applied Score plugin weight. 
+  FinalScore           int64
+}
+```
+
+#### The concept "ScenarioStep"
+
+ScenarioStep is: 
+- simply represented by numbers. like 1, 2, 3…
+- moved to next step **when it can no longer schedule any more Pods**.
+
+The following shows what happens at a single step in ScenarioStep:
+
+1. run all operations defined for that step.
+2. scheduler starts scheduling.
+3. scheduler stops scheduling when it can no longer schedule any more Pods.
+4. update status.scenarioResults.
+5. move to next step.
+
+##### Why scheduler needs to restarts/stops scheduling loop?
+
+To ensure that the results of the simulation do not vary significantly from run to run. 
+
+If we don't have "ScenarioStep" concept and when users want to define multiple operations for the same time, users expect them to run concurrently and them to be run at the same time. But, in practice it is difficult to run them at the same time, because the scheduler is constantly attempting to schedule. 
+For example, suppose a user has defined a scenario to create 1000 Nodes at the same time. Since it is strictly impossible to create 1000 Nodes at the same time, pending Pods will be scheduled to the Nodes created first. And depending on what order the Nodes were created, the results of the simulation may change.
+To prevent this, the scheduler needs to be stopped scheduling until 1000 Nodes are created.
+
+##### How to stop scheduling loop
+
+We can prevent a scheduling queue from releasing next Pods by replacing `Scheduler.NextPod` function.
+https://github.com/kubernetes/kubernetes/blob/867b5cc31b376c9f5d04cf9278112368b0337104/pkg/scheduler/scheduler.go#L75
+
+##### Adding events to running Scenario 
+
+It is allowed to add events while the Scenario is running.
+
+Note that it does not make sense to add past ScenarioStep events.
+The scenario will continue to run until all events in .spec.Events are completed, and when all events are completed, the scenario will be "Paused" phase in the case .spec.Events doesn't have "Done" operation.
+
+So, it is strongly recommended adding events to running Scenario only after Scenario have reached "Paused" phase. 
+(since ScenarioStep has stopped moving forward in "Paused" phase as described above.)
+Otherwise you may add the past ScenarioStep events and they are ignored by running Scenario.
+
+##### Configure when to update ScenarioResults
```

Comment:
```suggestion
##### Configure when to update ScenarioResult
```
Forgot to correct?
And others are `status.scenarioResults` → `status.scenarioResult`?

Reply:
Thanks, fixed.